File: c/luhn/_iterations/1/luhn.c
#include "luhn.h"

#include <ctype.h>
#include <string.h>

bool luhn(const char *num)
{
    int sum = 0;
    size_t digit_count = 0;
    for (const char *rnum = num + (strlen(num) - 1); rnum >= num; --rnum) {
        if (*rnum == ' ') {
            continue;
        }
        if (!isdigit((unsigned char) *rnum)) {
            return false;
        }

        int digit = *rnum - '0';
        if (++digit_count % 2 == 0) {
            digit *= 2;
            if (digit > 9) {
                digit -= 9;
            }
        }
        sum += digit;
    }

    return digit_count > 1 && sum % 10 == 0;
}

File: kotlin/luhn/_iterations/1/src/main/kotlin/Luhn.kt
object Luhn {
    fun isValid(str: String): Boolean {
        val strippedStr = str.filter { !it.isWhitespace() }
        return when {
            strippedStr.length <= 1 -> false
            strippedStr.any { !it.isDigit() } -> false
            else -> strippedStr.map { it - '0' }.isLuhn()
        }
    }

    private fun Iterable<Int>.isLuhn(): Boolean
        = this.reversed()
              .mapIndexed { idx, num -> if (idx % 2 != 0) num * 2 else num }
              .map { if (it > 9) it - 9 else it }
              .sum() % 10 == 0
}
